tests/layer/test_knowledge_graph.py

```diff
@@ -44,73 +44,23 @@ def triple_df(*values):
     return pd.DataFrame(values, columns=["source", "label", "target"])
 
 
-def test_complex(knowledge_graph):
-    # this test creates a random untrained model and predicts every possible edge in the graph, and
-    # compares that to a direct implementation of the scoring method in the paper
-    gen = KGTripleGenerator(knowledge_graph, 3)
-
-    # use a random initializer with a large positive range, so that any differences are obvious
-    init = initializers.RandomUniform(-1, 1)
-    complex_model = ComplEx(gen, 5, embeddings_initializer=init)
-    x_inp, x_out = complex_model.build()
-
-    model = Model(x_inp, x_out)
-    model.compile(loss=losses.BinaryCrossentropy(from_logits=True))
-
-    every_edge = itertools.product(
-        knowledge_graph.nodes(),
-        knowledge_graph._edges.types.pandas_index,
-        knowledge_graph.nodes(),
-    )
-    df = triple_df(*every_edge)
-
-    # check the model can be trained on a few (uneven) batches
-    model.fit(
-        gen.flow(df.iloc[:7], negative_samples=2),
-        validation_data=gen.flow(df.iloc[7:14], negative_samples=3),
-    )
-
-    # compute the exact values based on the model by extracting the embeddings for each element and
-    # doing the Re(<e_s, w_r, conj(e_o)>) inner product
-    s_idx = knowledge_graph._get_index_for_nodes(df.source)
-    r_idx = knowledge_graph._edges.types.to_iloc(df.label)
-    o_idx = knowledge_graph._get_index_for_nodes(df.target)
-
-    nodes, edge_types = complex_model.embeddings()
-    # the rows correspond to the embeddings for the given edge, so we can do bulk operations
-    e_s = nodes[s_idx, :]
-    w_r = edge_types[r_idx, :]
-    e_o = nodes[o_idx, :]
-    actual = (e_s * w_r * e_o.conj()).sum(axis=1).real
-
-    # predict every edge using the model
-    prediction = model.predict(gen.flow(df))
-
-    # (use an absolute tolerance to allow for catastrophic cancellation around very small values)
-    assert np.allclose(prediction[:, 0], actual, rtol=1e-3, atol=1e-14)
-
-    # the model is stateful (i.e. it holds the weights permanently) so the predictions with a second
-    # 'build' should be the same as the original one
-    model2 = Model(*complex_model.build())
-    prediction2 = model2.predict(gen.flow(df))
-    assert np.array_equal(prediction, prediction2)
-
-
-def test_complex_rankings():
+def _model_rankings_test(model_maker):
```

Comment:
could this be something like:

```suggestion
@pytest.mark.parametrize("model_maker", [ComplEx, DistMul])
def test_model_rankings(model_maker):
```
and remove `test_distmult_rankings` and `test_complex_rankings`

Reply:
I didn't go with that because some knowledge graph models may not take the same set of arguments (e.g. more than just the generator and embedding dimension), however, I think yours is probably better, overall. 👍